### ISSUE TITLE
Register a torch op for `functorch.experimental.control_flow.map` to lower

### ISSUE THREAD

**Issue by yeounoh** [pytorch/pytorch#113914]
### 🚀 The feature, motivation and pitch

We'd like to lower this `functorch.experimental.control_flow.map` and support in XLA. It's not registered as a `torch` op and gets unrolled. What would be the formal/best way to get support this?

Sample code,
```
import torch
from functorch.experimental import control_flow
from typing import List

def custom_backend(gm: torch.fx.GraphModule, example_inputs: List[torch.Tensor]):
    print("custom backend called with FX graph:")
    gm.graph.print_tabular()
    return gm.forward

torch._dynamo.reset()

def f(i, x):
  return x + i

def map1(x):
  return control_flow.map(f, torch.ones(5), x)

opt_map1 = torch.compile(map1, backend=custom_backend)

args = torch.ones(1, requires_grad=True) # (torch.randn((10, )), )
opt_map1(args)
```
fx.graph:
```
custom backend called with FX graph:
opcode         name        target                                                   args                      kwargs
-------------  ----------  -------------------------------------------------------  ------------------------  --------
placeholder    l_x_        L_x_                                                     ()                        {}
call_function  ones        <built-in method ones of type object at 0x7f6243a6bc00>  (5,)                      {}
call_method    getitem     __getitem__                                              (ones, 0)                 {}
get_attr       map_body_0  map_body_0                                               ()                        {}
call_function  map_1       map                                                      (map_body_0, ones, l_x_)  {}
output         output      output    
```

Dumped IR (unrolled)
```
IR {
  %0 = f32[] prim::Constant(), location=__call__@_ops.py:516, xla_shape=f32[], value=1
  %1 = f32[] prim::Constant(), location=__call__@_ops.py:516, xla_shape=f32[], value=1
  %2 = f32[] aten::mul(%1, %0), location=__call__@_ops.py:516, xla_shape=f32[]
  %3 = f32[1]{0} xla::device_data(), location=<module>@torchxla_control_flow_map.py:20, xla_shape=f32[1]{0}, device=TPU:0
  %4 = f32[1]{0} aten::add(%3, %2), location=__call__@_ops.py:516, xla_shape=f32[1]{0}
  %5 = f32[] prim::Constant(), location=__call__@_ops.py:516, xla_shape=f32[], value=1
  %6 = f32[] prim::Constant(), location=__call__@_ops.py:516, xla_shape=f32[], value=1
  %7 = f32[] aten::mul(%6, %5), location=__call__@_ops.py:516, xla_shape=f32[]
  %8 = f32[1]{0} aten::add(%3, %7), location=__call__@_ops.py:516, xla_shape=f32[1]{0}
  %9 = f32[] prim::Constant(), location=__call__@_ops.py:516, xla_shape=f32[], value=1
  %10 = f32[] prim::Constant(), location=__call__@_ops.py:516, xla_shape=f32[], value=1
  %11 = f32[] aten::mul(%10, %9), location=__call__@_ops.py:516, xla_shape=f32[]
  %12 = f32[1]{0} aten::add(%3, %11), location=__call__@_ops.py:516, xla_shape=f32[1]{0}
  %13 = f32[] prim::Constant(), location=__call__@_ops.py:516, xla_shape=f32[], value=1
  %14 = f32[] prim::Constant(), location=__call__@_ops.py:516, xla_shape=f32[], value=1
  %15 = f32[] aten::mul(%14, %13), location=__call__@_ops.py:516, xla_shape=f32[]
  %16 = f32[1]{0} aten::add(%3, %15), location=__call__@_ops.py:516, xla_shape=f32[1]{0}
  %17 = f32[] prim::Constant(), location=__call__@_ops.py:516, xla_shape=f32[], value=1
  %18 = f32[] prim::Constant(), location=__call__@_ops.py:516, xla_shape=f32[], value=1
  %19 = f32[] aten::mul(%18, %17), location=__call__@_ops.py:516, xla_shape=f32[]
  %20 = f32[1]{0} aten::add(%3, %19), location=__call__@_ops.py:516, xla_shape=f32[1]{0}
  %21 = f32[5,1]{1,0} aten::stack(%20, %16, %12, %8, %4), location=_stack_pytree@_map.py:300, xla_shape=f32[5,1]{1,0}, dim=0, ROOT=0
}
```

cc @JackCaoG @ezyang @brianhir

### Alternatives

We can try to work with the fx.graph, but feels like more of a hack.

**Comment by ezyang** [comment]
cc @ydwu4 @zou3519

**Comment by miladm** [comment]
Following up on this issue - is there a plan of record for supporting this feature in Q1?
@ydwu4 @zou3519

**Comment by zou3519** [comment]
Are you looking to lower a map() or something more like for_i_loop or while_loop? It's possible that we'll implement map in a way that it desugars into one of the others.

**Comment by ydwu4** [comment]
Sorry for the late reply.  For the given example, the map operator is preserved as torch.ops.higher_order.map_impl in the fx.Graph and seems get de-sugared at some time later (is it at XLA dispatch key? I lack some context here..). What's the lowered target you're expecting?

BTW, I'm doing a major refactoring for the map_impl operators schema to make it normalized (remove the num_mapped and replace positional args with two tuples, mapped and operands) and change its name to (torch.ops.higher_order.map) to make it consistent with `torch.cond` and `torch.ops.higher_order.cond`. These changes will land soon (hopefully).

**Comment by ManfeiBai** [comment]
we want to something like `fori_loop` which could be lowered with `while`, so `torch.scan` is the first choice, and `torch.map`/`torch.fori_loop` are avaiable to meet the requirement too

**Comment by yeounoh** [comment]
> Are you looking to lower a map() or something more like for_i_loop or while_loop? It's possible that we'll implement map in a way that it desugars into one of the others.

Hi @zou3519 thank you for pointing it out -- yea we want a for/while loop op. I am also following the other issue, that requests scanned, which would do as well.

**Comment by yeounoh** [comment]
> Sorry for the late reply. For the given example, the map operator is preserved as torch.ops.higher_order.map_impl in the fx.Graph and seems get de-sugared at some time later (is it at XLA dispatch key? I lack some context here..). What's the lowered target you're expecting?
> 
> BTW, I'm doing a major refactoring for the map_impl operators schema to make it normalized (remove the num_mapped and replace positional args with two tuples, mapped and operands) and change its name to (torch.ops.higher_order.map) to make it consistent with `torch.cond` and `torch.ops.higher_order.cond`. These changes will land soon (hopefully).

Hi @ydwu4 thanks for the update. Based on the point @zou3519 raised in https://github.com/pytorch/pytorch/issues/113914#issuecomment-1883288129, once we have a fori_loop op, then we can lower it to XLA's while op (target). I am guessing that the `map` is being unrolled after Dynamo maybe because it's implemented with `for` somehwere and we haven't directly lowered `map`, yet?

**Comment by ydwu4** [comment]
> maybe because it's implemented with for somewhere

Yes, see map's CompositeExplicitAutograd key implementation here: https://github.com/pytorch/pytorch/blob/main/torch/_higher_order_ops/map.py#L313-L320. 

I'm wondering how the Dumped IR (unrolled) shown in the issue description is generated. Is it generated by some kind of XLA's tracer? I'm totally new with the lowering stack from PyTorch to XLA... Do you mind sharing a bit more on it?

**Comment by ydwu4** [comment]
Thanks for the repro but somehow my environment is not set up properly with some missing shared library error. I'm interested in a quick intro to how torch_xlas integrates with pytorch. Is the current problem that we should implement map's XLA dispatcher key by leveraging xla's primitives while_loop or something similar? @zou3519
